halo2_middleware/src/expression.rs

```diff
@@ -0,0 +1,287 @@
+use core::cmp::max;
+use core::ops::{Add, Mul, Neg, Sub};
+use ff::Field;
+use std::iter::{Product, Sum};
+
+pub trait Variable: Clone + Copy + std::fmt::Debug + Eq + PartialEq {
+    /// Degree that an expression would have if it was only this variable.
+    fn degree(&self) -> usize;
+
+    /// Return true if this variable is a virtual abstraction that may be transformed later.  Two
+    /// expressions containing virtual variables cannot be added or multiplied, this means an
+    /// expression can only contain up to one virtual variable.  An expression containing a virtual
+    /// variable cannot appear in a lookup.  Originally this is only fulfilled by the simple
+    /// selector abstraction in the halo2 frontend.
+    /// Overwrite this if the Variable has a virtual case.
+    fn is_virtual(&self) -> bool {
+        false
+    }
+
+    /// Return the name of the virtual variable case.  Originally this is used for "simple
```

Comment:
Originally this was a kind of a "hack" to keep a particular behaviour that the frontend Expression needs: adding/multiplying two expressions with simple selectors is forbidden.  Since I wanted to make the Expression generic, I was thinking on a way in which Expressions with selectors can have some special rules, and my proposal was to call the simple selector a "virtual variable", in the sense that it doesn't map 1 to 1 with a plonkish constraint variable.  And then implementations of the Variable type must define weather there is such a case or not.  Then the `virtual_var_case` method returns a string to be used in the panic message, so that instead of getting `attempted to multiply two expressions containing virtual variables` we get `attempted to multiply two expressions containing selectors`.

On the other hand, in this PR the frontend is using a different Expression type in order to not break the API, so there's no case for the generic Expression that needs support for this "virtual variable".  So I think I'll just remove this and simplify the code.

Reply:
I've removed all the "virtual variable" code in https://github.com/privacy-scaling-explorations/halo2/pull/290/commits/f3dc8046317f5dbc39b37ae50736595e9f21acc5